### PR TITLE
fix(dsg): [BREAKING CHANGE] duplicated modules 

### PR DESCRIPTION
**Pull request by overbit**
Close: #5791 #5837
## PR Details

In order to fix the duplicated modules in the modules lists, this PR brings a breaking change to the way we interact with file generation by updating the type to a map type
This makes it clear to the developer if a file is going to be overwritten or not and it will allow dsg to always write one file only one time during a single run.

⚠️ This PR contains breaking changes and it will need to go live with https://github.com/amplication/plugins/pull/113 

**As part of this PR, I've published `@amplication/code-gen-types@2.0.0` with `beta` tag. Once released we need to just retag this version with `latest` tag.**

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-code-gen-types/src/code-gen-types.ts [hunk 1]
```diff
@@ -129,6 +124,77 @@ export type Module = {
...
+  async replaceModulesCode(fn: (code: string) => string): Promise<void> {
+    for await (const module of Object.values(this.map)) {
+      module.code = fn(module.code);
+      this[module.path] = module;
+    }
+  }
+
+  /**
+   * Returns all modules in the map
+   * @returns An array of modules
```
Comment (abrl91) [inline packages/amplication-code-gen-types/src/code-gen-types.ts:191]:
```suggestion
   * @returns An array of modules
```

File: packages/data-service-generator/src/prepare-context.ts [hunk 2]
```diff
@@ -26,7 +25,7 @@ export async function prepareContext(
   dSGResourceData: DSGResourceData,
   logger: ILogger,
   pluginInstallationPath?: string
-): Promise<Module[]> {
+): Promise<void> {
```
Comment (abrl91) [inline packages/data-service-generator/src/prepare-context.ts:28]: Why did you change the type?

File: packages/data-service-generator/src/prepare-context.ts [hunk 3]
```diff
@@ -68,7 +67,6 @@ export async function prepareContext(
   );
 
   context.plugins = plugins;
-  return [];
```
Comment (abrl91) [inline packages/data-service-generator/src/prepare-context.ts:71]: why did you remove that?

File: packages/data-service-generator/src/prepare-context.ts (hunk 2 above)
Comment (overbit) [reply packages/data-service-generator/src/prepare-context.ts:28]: To reflect the actual behaviour. It never returns Module[]

File: packages/data-service-generator/src/prepare-context.ts (hunk 3 above)
Comment (overbit) [reply packages/data-service-generator/src/prepare-context.ts:71]: It is never called with the intent of creating a Module array

File: packages/amplication-code-gen-types/src/code-gen-types.ts [hunk 4]
```diff
@@ -129,6 +124,76 @@ export type Module = {
...
+      await this.merge(map);
+    }
+  }
+
+  async set(path: string, module: Module) {
+    if (this.map[path]) {
+      await this.logger.warn(`Module ${path} already exists. Overwriting...`);
+    }
+    this.map[path] = module;
+  }
```
Comment (overbit) [inline packages/amplication-code-gen-types/src/code-gen-types.ts:151]: Auto review for @overbit. How about changing the signature of set to accept only `Module`?

```suggestion
  async set(module: Module) {
    if (this.map[module.path]) {
      await this.logger.warn(`Module ${module.path} already exists. Overwriting...`);
    }
    this.map[module.path] = module;
  }
```

**Comment by mulygottlieb** [comment]
@overbit @barshimi please complete the review and agree on the changes still needed to merge this. I want this done ASAP.

File: packages/amplication-code-gen-types/src/code-gen-types.ts (hunk 4 above)
Comment (yuval-hazaz) [reply packages/amplication-code-gen-types/src/code-gen-types.ts:151]: @overbit I Agree. let's use only module

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
